backend/azure/dtmb/dtmb.go

```diff
@@ -0,0 +1,531 @@
+package dtmb
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/microsoft/durabletask-go/api"
+	"github.com/microsoft/durabletask-go/backend"
+	"github.com/microsoft/durabletask-go/internal/protos"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	dtmbprotos "github.com/microsoft/durabletask-go/backend/azure/dtmb/internal/backend/v1"
+	"github.com/microsoft/durabletask-go/backend/azure/dtmb/internal/utils"
+)
+
+const (
+	// DefaultEndpoint is the default endpoint for the DTMB service.
+	defaultEndpoint = "localhost:50051"
+)
+
+type dtmb struct {
+	logger                    backend.Logger
+	endpoint                  string
+	options                   *DTMBOptions
+	orchestrationQueue        utils.SyncQueue[dtmbprotos.ExecuteOrchestrationMessage]
+	activityQueue             utils.SyncQueue[dtmbprotos.ExecuteActivityMessage]
+	orchestrationHistoryCache utils.OrchestrationHistoryCache
+	connectWorkerClientStream chan *dtmbprotos.ConnectWorkerClientMessage
+	clientClient              dtmbprotos.TaskHubClientClient
+	workerClient              dtmbprotos.TaskHubWorkerClient
+	workerCancelFunc          context.CancelFunc
+}
+
+type DTMBOptions struct {
+	Endpoint string
+}
+
+func NewDTMBOptions(endpoint string) *DTMBOptions {
+	if endpoint != "" {
+		return &DTMBOptions{
+			Endpoint: endpoint,
+		}
+	} else {
+		return &DTMBOptions{
+			Endpoint: defaultEndpoint,
+		}
+	}
+}
+
+func NewDTMB(opts *DTMBOptions, logger backend.Logger) (*dtmb, error) {
+	be := &dtmb{
+		logger: logger,
+	}
+
+	if opts == nil {
+		opts = NewDTMBOptions(defaultEndpoint)
+	}
+	be.options = opts
+	be.endpoint = opts.Endpoint
+
+	// The following queues are used to store messages received from the server
+	be.orchestrationQueue = utils.NewSyncQueue[dtmbprotos.ExecuteOrchestrationMessage]()
+	be.activityQueue = utils.NewSyncQueue[dtmbprotos.ExecuteActivityMessage]()
+
+	be.orchestrationHistoryCache = utils.NewOrchestrationHistoryCache(nil) // TODO: make capacity configurable
+
+	ctx := context.Background()
+	connCtx, connCancel := context.WithTimeout(ctx, 15*time.Second) // TODO: make this a configurable timeout
```

Comment:
Should probably change the signature of `NewDTMB` to accept a context from the user, so the user can set a timeout

Reply:
There's also a context associated with the app's lifecycle. It's common pattern to have a context started when the app starts, and cancel it when the app is shutting down gracefully

Reply:
@cgillum that's what my comment is saying - make the timeout configurable in DTMBOptions.

But anyways, I was not looking for a review on those details yet - I will be refactoring some of the options anyway. I am worried about the data mapping for now.